### ISSUE TITLE
bogus diagnostic `deduction_guide_bad_trailing_return_type`

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#54909]
The following code is passed in gcc and msvc, but clang gives a bogus diagnostic `deduced type 'B<T>' (aka 'A<type-parameter-0-0>') of deduction guide is not written as a specialization of template 'A'`.

The reason is that when clang compares the equality of two Template-kind `TemplateName`, it doesn't seem to consider the `TypeAliasTemplateDecl` case. 

```
template<typename T> struct A {};
A(void) -> A<int>;

template<typename T> using B = A<T>; 

template<typename T> A(T) -> B<T>; 
```

**Comment by Bekenn** [comment]
The prior behavior was correct; this change should be reverted.

Relevant excerpt from the standard:

> &nbsp;&nbsp;&nbsp;&nbsp;[*deduction-guide*:](https://eel.is/c++draft/temp.deduct.guide#nt:deduction-guide)
> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;[*explicit-specifier*](https://eel.is/c++draft/dcl.fct.spec#nt:explicit-specifier)<sub>opt</sub> [*template-name*](https://eel.is/c++draft/temp.names#nt:template-name) `(` [*parameter-declaration-clause*](https://eel.is/c++draft/dcl.fct#nt:parameter-declaration-clause) `)` [*requires-clause*](https://eel.is/c++draft/temp.pre#nt:requires-clause)<sub>opt</sub> `->` [*simple-template-id*](https://eel.is/c++draft/temp.names#nt:simple-template-id) `;`

[[temp.deduct.guide]/3](https://eel.is/c++draft/temp.deduct.guide#3):

> The [*simple-template-id*](https://eel.is/c++draft/temp.names#nt:simple-template-id) shall name a class template specialization.

This is saying that the *template-name* in the *simple-template-id* in the return type shall be the actual name of the class template rather than some equivalent alias template.  To make it even more clear:

> The [*template-name*](https://eel.is/c++draft/temp.names#nt:template-name) shall be the same [*identifier*](https://eel.is/c++draft/lex.name#nt:identifier) as the [*template-name*](https://eel.is/c++draft/temp.names#nt:template-name) of the [*simple-template-id*](https://eel.is/c++draft/temp.names#nt:simple-template-id).

This is saying that the name of the deduction guide must match the name of the template in the return type.

**Comment by a-tarasyuk** [comment]
@Bekenn thanks for highlighting that detail in the spec. @hokein @erichkeane I've created PR to revert these changes #118165.

**Comment by Bekenn** [comment]
Thanks for addressing this!

**Comment by hokein** [comment]
Thanks for pointing this out.

I should have checked the standard before filing the bug. Based on the standard's specifications, I think the previous behavior was correct, and we should revert the change.
